src/master-playlist-controller.js

```diff
@@ -0,0 +1,519 @@
+
+    // check that everything is ready to begin buffering
+
+    // 1) the video is a live stream of unknown duration
+    if (this.hlsHandler.duration() === Infinity &&
+
+        // 2) the player has not played before and is not paused
+        this.hlsHandler.tech_.played().length === 0 &&
+        !this.hlsHandler.tech_.paused() &&
+
+        // 3) the active media playlist is available
+        media) {
+
+      this.load();
+
+      // 4) the video element or flash player is in a readyState of
+      // at least HAVE_FUTURE_DATA
+      if (this.hlsHandler.tech_.readyState() >= 1) {
+
+        // trigger the playlist loader to start "expired time"-tracking
+        this.masterPlaylistLoader_.trigger('firstplay');
+
+        // seek to the latest media position for live videos
+        seekable = this.hlsHandler.seekable();
+        if (seekable.length) {
+          this.hlsHandler.tech_.setCurrentTime(seekable.end(0));
+        }
+      }
+    }
+  }
+
+  handleSourceOpen_() {
+    // if autoplay is enabled, begin playback. This is duplicative of
+    // code in video.js but is required because play() must be invoked
+    // *after* the media source has opened.
+    if (this.hlsHandler.tech_.autoplay()) {
+      this.hlsHandler.play();
+    }
+  }
+
+  /*
+   * Blacklists a playlist when an error occurs for a set amount of time
+   * making it unavailable for selection by the rendition selection algorithm
+   * and then forces a new playlist (rendition) selection.
+   */
+  blacklistCurrentPlaylist_(error) {
+    let currentPlaylist;
+    let nextPlaylist;
+
+    // If the `error` was generated by the playlist loader, it will contain
+    // the playlist we were trying to load (but failed) and that should be
+    // blacklisted instead of the currently selected playlist which is likely
+    // out-of-date in this scenario
+    currentPlaylist = error.playlist || this.masterPlaylistLoader_.media();
+
+    // If there is no current playlist, then an error occurred while we were
+    // trying to load the master OR while we were disposing of the tech
+    if (!currentPlaylist) {
+      this.hlsHandler.error = error;
+      return this.mediaSource.endOfStream('network');
+    }
+
+    // Blacklist this playlist
+    currentPlaylist.excludeUntil = Date.now() + BLACKLIST_DURATION;
+
+    // Select a new playlist
+    nextPlaylist = this.hlsHandler.selectPlaylist();
+
+    if (nextPlaylist) {
+      videojs.log.warn('Problem encountered with the current ' +
+                       'HLS playlist. Switching to another playlist.');
+
+      return this.masterPlaylistLoader_.media(nextPlaylist);
+    }
+    videojs.log.warn('Problem encountered with the current ' +
+                     'HLS playlist. No suitable alternatives found.');
+    // We have no more playlists we can select so we must fail
+    this.hlsHandler.error = error;
+    return this.mediaSource.endOfStream('network');
+  }
+
+  pause() {
```

Comment:
The name of this method might be confusing. You wouldn't want to pause the segment loaders if the player is paused.